lesson-3-sheduling-all-around/README.md

```diff
@@ -0,0 +1,266 @@
+
+Scheduling
+==========
+
+default scheduler is described at: https://kubernetes.io/docs/concepts/scheduling-eviction/kube-scheduler/
+
+
+The scheduler is working in multiple phases
+1) finding a feasible se of nodes 
+1) scoring feasible nodes so we can decide where workload will be places
+
+Why this is so important? Because it can make your application not behave correctly and lead to potential downtime. Imagine situations like:
+
+1) all application pods running on same node
+1) all application pods running on node on same physical host
+1) pods which should be together are spread on multiple nodes
+1) ...
+
+We can prevent this by concept likes affinities and anti-affinities. They are described in a docs at: https://docs.openshift.com/container-platform/4.10/nodes/scheduling/nodes-scheduler-node-affinity.html
+
+
+As an example to practice this we will create a three pods: 
+- p1 - which defines "security" to value "S1"
+
+``` yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: p1
+  labels:
+    security: S1
+spec:
+  containers:
+  - name: security-s1
+    image: docker.io/ocpqe/hello-pod
+```
+
+- p2 - which has affinity to be run on same node as "p1" using the selector be scheduled on same node as the pods with label "security" set to "S1"
+
+
+``` yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: p2
+  labels:
+    security: S1
+spec:
+  containers:
+  - name: security-s1
+    image: docker.io/ocpqe/hello-pod
+    podAffinity:
+      requiredDuringSchedulingIgnoredDuringExecution:
+      - labelSelector:
+          matchExpressions:
+          - key: security
+            operator: In
+            values:
+            - S1
+```
+
+- p2 - which has antiAffinity to NOT be run on same node as "p1" using the selector be scheduled on same node as the pods with label "security" set to "S1"
+
+``` yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: p3
+spec:
+  affinity:
+    podAntiAffinity:
+      requiredDuringSchedulingIgnoredDuringExecution:
+      - labelSelector:
+          matchExpressions:
+          - key: security
+            operator: In
+            values:
+            - S1
+        topologyKey: kubernetes.io/hostname
+  containers:
+  - name: pod-antiaffinity
+    image: docker.io/ocpqe/hello-pod
+```
+ 
+Now you can use `oc` command to double check if everything is running correctly.
+
+``` bash
+oc get pods -o jsonpath='{range .items[*]}{.metadata.name}{"\t"}{.spec.nodeName}{"\n"}'
+```
+
+As a task for yourself please create pod "p4" which will be scheduled on same node as pod "p3".
```

Comment:
nice, that's a good task for them

I hope we'll have multiple nodes in our cluster :D